src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -816,37 +820,84 @@ public static (RegexNode LoopNode, (char Char, string? String, char[]? Chars) Li
+                // characters that either are ASCII or that don't participate in case conversion).
+                Debug.Assert(
+                    !RegexCharClass.ParticipatesInCaseConversion(ordinalCaseInsensitivePrefix[0]) ||
+                    ordinalCaseInsensitivePrefix[0] < 128);
+
+                if (RegexCharClass.ParticipatesInCaseConversion(ordinalCaseInsensitivePrefix[0]))
+                {
+                    if (RegexCharClass.CharInClass((char)(ordinalCaseInsensitivePrefix[0] | 0x20), firstChild.Str!) ||
+                        RegexCharClass.CharInClass((char)(ordinalCaseInsensitivePrefix[0] & ~0x20), firstChild.Str!))
+                    {
+                        return null;
+                    }
+                }
+                else if (RegexCharClass.CharInClass(ordinalCaseInsensitivePrefix[0], firstChild.Str!))
+                {
+                    return null;
+                }
 
-                case RegexNodeKind.Multi when !RegexCharClass.CharInClass(nextChild.Str![0], firstChild.Str!):
-                    return (firstChild, ('\0', nextChild.Str, null));
+                return (firstChild, ('\0', ordinalCaseInsensitivePrefix, StringComparison.OrdinalIgnoreCase, null));
+            }
+
+            // Is the set loop followed by a set we can search for? Whereas the above helpers will drill down into
+            // children as is appropriate, to examine a set here, we need to drill in ourselves. We can drill through
+            // atomic and capture nodes, as they don't affect flow control, and into the left-most node of a concatenate,
+            // as the first child is guaranteed next. We can also drill into a loop or lazy loop that has a guaranteed
+            // iteration, for the same reason as with concatenate.
+            while ((nextChild.Kind is RegexNodeKind.Atomic or RegexNodeKind.Capture or RegexNodeKind.Concatenate) ||
+                   (nextChild.Kind is RegexNodeKind.Loop or RegexNodeKind.Lazyloop && nextChild.M >= 1))
+            {
+                nextChild = nextChild.Child(0);
+            }
 
-                case RegexNodeKind.Set when !RegexCharClass.IsNegated(nextChild.Str!):
-                    Span<char> chars = stackalloc char[5]; // maximum number of chars optimized by IndexOfAny
-                    chars = chars.Slice(0, RegexCharClass.GetSetChars(nextChild.Str!, chars));
-                    if (!chars.IsEmpty)
+            // If the resulting node is a set with at least one iteration, we can search for it.
+            if (nextChild.IsSetFamily &&
+                !RegexCharClass.IsNegated(nextChild.Str!) &&
+                (nextChild.Kind is RegexNodeKind.Set || nextChild.M >= 1))
+            {
+                Span<char> chars = stackalloc char[5]; // maximum number of chars optimized by IndexOfAny
+                chars = chars.Slice(0, RegexCharClass.GetSetChars(nextChild.Str!, chars));
+                if (!chars.IsEmpty)
```

Comment:
is the empty test here a runtime improvement vs just letting the foreach skip?

it seems to me it just causes the empty check twice:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYabi6kACYlVgQQAoBlAA7YAdgB4wAC2xQAfK0nTcAShoBvGq02sAlgDNWPAIQKouAHQBJXAFF8AjAE8V1La3UvXW3dBjZJBk3kdYXkpUyVNDU9Nd2jogGEIYVwIABsYMwB1KG0MGB4wJQBuKOiAX1LWCupqphZ2TgAhfiExEzkTZTVK71g/CQCwoO0QzojWSti4xOS0jOzc/MKSj01q2tXWIA===

```asm
    L0007: mov esi, [rcx+8]
    L000a: test esi, esi
    L000c: je short L002f
    L000e: mov rdi, [rcx]
    L0011: xor ebx, ebx
    L0013: test esi, esi
    L0015: jle short L002f
```
vs 
```asm
    L0007: mov rsi, [rcx]
    L000a: mov edi, [rcx+8]
    L000d: xor ebx, ebx
    L000f: test edi, edi
    L0011: jle short L002b
```

Reply:
maybe a potential JIT improvement.

Reply:
> is the empty test here a runtime improvement vs just letting the foreach skip?

It's not an efficiency thing: it's necessary for correctness. If there are too many chars in the set to fit into the specified chars span, GetSetChars returns 0, which means we can't trust the contents of chars.

Reply:
I guess IsEmpty could move down to just around https://github.com/dotnet/runtime/pull/93190/files/6aad45619b9884cd7cd379f4bdd15564a16f8d4b#diff-3a7a4260e3a2cd96d2beb16ed4ec8ac508678b139c67f70a37f81be6433cf610R899
then if the foreach ends up returning IsEmpty is never called.

Reply:
I don't understand... move down to where?  Are you saying put the foreach loop above the IsEmpty check? If so, what benefit does that provide?

Reply:
yes .well, not a meaningful one, but in the case chars is not empty, and it goes into the foreach, and returns from the method from within that, then this IsEmpty check wouldn't execute, whereas right now it does either way. am I spacing?

Reply:
> but in the case chars is not empty, and it goes into the foreach, and returns from the method from within that, then this IsEmpty check wouldn't execute, whereas right now it does either way

And if the foreach loop moved earlier, then you're checking its count and if it's 0 then you're checking IsEmpty, so with both orderings in one case you're checking twice and in the other case once.

Reply:
Yeah in the current case if empty there is one check, otherwise two. With this change, if non empty and returning from within loop one check, otherwise two. I guess I was implicitly assuming non empty was more common, for no reason.

Anyway this matters zero, I was just curious what mistake I was making...

Reply:
https://github.com/dotnet/runtime/issues/93708